File: python/src/swmmrs/objects/_collections.py
"""Shared mapping adapters for configured SWMM object views."""

from __future__ import annotations as _annotations

from collections.abc import Iterable as _Iterable, Iterator as _Iterator, Mapping as _Mapping
from typing import (
    TYPE_CHECKING as _TYPE_CHECKING,
    Generic as _Generic,
    Never as _Never,
    SupportsIndex as _SupportsIndex,
    cast as _cast,
    final as _final,
    overload as _overload,
)

from ._base import _COLLECTION_TOKEN, _T, _ObjectFamily

if _TYPE_CHECKING:
    from .. import Simulation
    from ._links import Link, LinkCollection
    from ._nodes import Node, NodeCollection
    from ._subcatchments import Subcatchment, SubcatchmentCollection


class _Collection(_Mapping[str, _T], _Generic[_T]):
    __slots__ = ("_family", "_generation", "_simulation")

    def __init__(
        self,
        simulation: Simulation,
        generation: int,
        family: _ObjectFamily,
        token: object,
    ) -> None:
        if token is not _COLLECTION_TOKEN:
            raise TypeError("collections are created by Simulation")
        self._simulation = simulation
        self._generation = generation
        self._family = family

    def __getitem__(self, key: str | int) -> _T:
        """Return an object view by case-insensitive configured ID.

        Parameters
        ----------
        key
            Object ID as a string or non-boolean integer.

        Returns
        -------
        _T
            Generation-bound object view.

        Raises
        ------
        KeyError
            If no configured object has the requested ID.
        TypeError
            If `key` is not a string or non-boolean integer.
        """
        return _cast(
            _T,
            self._simulation._owner.collection_view(
                self._simulation, self._generation, self._family, key
            ),
        )

    def by_index(self, index: int) -> _T:
        """Return an object view by its zero-based configured position.

        Parameters
        ----------
        index
            Zero-based position in configured project order.

        Returns
        -------
        _T
            Generation-bound object view.

        Raises
        ------
        IndexError
            If `index` is outside the configured collection.
        TypeError
            If `index` is not a non-boolean integer.
        """
        return _cast(
            _T,
            self._simulation._owner.collection_view_at(
                self._simulation, self._generation, self._family, index
            ),
        )

    def __iter__(self) -> _Iterator[str]:
        """Return an iterator over object IDs in configured project order."""
        return iter(self._simulation._owner.collection_ids(self._generation, self._family))

    def __len__(self) -> int:
        """Return the number of configured objects in this collection."""
        return self._simulation._owner.collection_count(self._generation, self._family)

    def __repr__(self) -> str:
        identifiers = self._simulation._owner.collection_ids(self._generation, self._family)
        return f"{type(self).__name__}({identifiers!r})"

    def __contains__(self, key: object) -> bool:
        """Return whether a case-insensitive object ID is configured.

        Parameters
        ----------
        key
            Candidate object ID.

        Returns
        -------
        bool
            `True` when `key` identifies a configured object; otherwise `False`.
        """
        return self._simulation._owner.collection_contains(self._generation, self._family, key)

    def __copy__(self) -> _Never:
        raise TypeError(f"{type(self).__name__} cannot be copied")

    def __deepcopy__(self, memo: object) -> _Never:
        raise TypeError(f"{type(self).__name__} cannot be deep-copied")

    def __reduce__(self) -> _Never:
        raise TypeError(f"{type(self).__name__} cannot be pickled")

    def __reduce_ex__(self, protocol: _SupportsIndex) -> _Never:
        raise TypeError(f"{type(self).__name__} cannot be pickled")


@_final
class ObjectCollection(_Collection[_T], _Generic[_T]):
    """Provide mapping-style access to configured project objects.

    Notes
    -----
    Object IDs are case-insensitive. Use `by_index()` for configured order.
    """

    __slots__ = ()


@_overload
def _new_collection(
    simulation: Simulation,
    generation: int,
    family: _ObjectFamily,
    collection_type: type[ObjectCollection[_T]],
) -> ObjectCollection[_T]: ...


@_overload
def _new_collection(
    simulation: Simulation,
    generation: int,
    family: _ObjectFamily,
    collection_type: type[NodeCollection],
) -> NodeCollection: ...


@_overload
def _new_collection(
    simulation: Simulation,
    generation: int,
    family: _ObjectFamily,
    collection_type: type[LinkCollection],
) -> LinkCollection: ...


@_overload
def _new_collection(
    simulation: Simulation,
    generation: int,
    family: _ObjectFamily,
    collection_type: type[SubcatchmentCollection],
) -> SubcatchmentCollection: ...


def _new_collection(
    simulation: Simulation,
    generation: int,
    family: _ObjectFamily,
    collection_type: type[_Collection[_T]],
) -> _Collection[_T]:
    return collection_type(simulation, generation, family, _COLLECTION_TOKEN)

File: js/src/swmmrs/objects/collections.ts
import { ObjectNotFoundError } from "../exceptions.js";
import type { Call } from "../protocol.js";
import type { NodeSnapshot, LinkSnapshot, SubcatchmentSnapshot } from "../snapshots.js";
import { Node, type NodeQualitySnapshot, type NodeStatisticsSnapshot } from "./nodes.js";
import { Link, type LinkQualitySnapshot, type LinkStatisticsSnapshot } from "./links.js";
import { Subcatchment, RainGage, type SubcatchmentQualitySnapshot, type SubcatchmentStatisticsSnapshot } from "./subcatchments.js";

/**
 * Configured-order handles with synchronous, case-insensitive ID lookup and stable identity.
 * Obtain simulation collections from the owner rather than constructing a replacement view.
 * All access checks the owner lifecycle; no access is valid after close.
 * @typeParam T - Handle type returned by lookup and iteration.
 */
export class ObjectCollection<T> implements Iterable<T> {
  readonly #ids: readonly string[];
  readonly #byId: Map<string, string>;
  readonly #views = new Map<string, T>();
  readonly #create: (id: string) => T;
  readonly #assertOpen: () => void;

  /** Construct the shared collection implementation.
   * @param ids - Canonical IDs in configured order; copied on construction.
   * @param create - Factory called once per looked-up canonical ID.
   * @param assertOpen - Lifecycle guard called before collection access.
   */
  constructor(ids: readonly string[], create: (id: string) => T, assertOpen: () => void) {
    this.#ids = Object.freeze([...ids]);
    this.#byId = new Map(ids.map((id) => [id.toLowerCase(), id]));
    this.#create = create;
    this.#assertOpen = assertOpen;
  }
  /** Canonical IDs in configured order; a frozen array. */
  get ids(): readonly string[] { this.#assertOpen(); return this.#ids; }
  /** Number of configured objects. */
  get size(): number { this.#assertOpen(); return this.#ids.length; }
  /** Test membership synchronously.
   * @param id - Case-insensitive object ID.
   * @returns Whether the ID is configured.
   */
  has(id: string): boolean { this.#assertOpen(); return this.#byId.has(id.toLowerCase()); }
  /** Look up a stable handle synchronously.
   * @param id - Case-insensitive object ID.
   * @returns The same handle instance for repeated lookups of one canonical ID.
   * @throws ObjectNotFoundError for an unknown ID; lifecycle errors after close.
   */
  get(id: string): T {
    this.#assertOpen();
    const canonical = this.#byId.get(id.toLowerCase());
    if (canonical === undefined) throw new ObjectNotFoundError({ message: `Unknown object: ${id}`, code: 505, operation: "lookup" });
    let view = this.#views.get(canonical);
    if (view === undefined) { view = this.#create(canonical); this.#views.set(canonical, view); }
    return view;
  }
  /** Look up a configured-order handle synchronously.
   * @param index - Zero-based integer index; negative and fractional values are invalid.
   * @returns The stable handle at the index.
   * @throws RangeError for an out-of-range or noninteger index.
   */
  at(index: number): T {
    this.#assertOpen();
    const id = this.#ids[index];
    if (!Number.isInteger(index) || id === undefined) throw new RangeError(`Object index out of range: ${index}`);
    return this.get(id);
  }
  /** Iterate handles in configured order.
   * @returns Synchronous iterator, checking the owner on each lookup.
   */
  *[Symbol.iterator](): IterableIterator<T> {
    this.#assertOpen();
    for (const id of this.#ids) yield this.get(id);
  }
}

/** Owner-bound node handles. Snapshot selections preserve requested order; omit IDs for all nodes. */
export class NodeCollection extends ObjectCollection<Node> {
  readonly #call: Call;
  private constructor(ids: readonly string[], call: Call, assertOpen: () => void) {
    super(ids, (id) => Node.create(id, call), assertOpen); this.#call = call;
  }
  /** @internal */
  static create(ids: readonly string[], call: Call, assertOpen: () => void): NodeCollection { return new NodeCollection(ids, call, assertOpen); }
  /** Read node hydraulics in `running`, `complete`, or `ended`.
   * @param ids - Ordered selection; omit for all, use [] for empty columns. Unknown or case-insensitive duplicate IDs reject.
   * @returns Detached columns aligned with canonical `objectIds`, without a timestamp.
   */
  snapshot(ids?: readonly string[]): Promise<NodeSnapshot> { return this.#call("nodes", ids); }
  /** Read node quality in `running`, `complete`, or `ended`.
   * @param ids - Ordered IDs; omit for all or use [] for none. Unknown/duplicate IDs reject.
   * @returns Detached pollutant-major matrices; inner rows follow `objectIds`.
   */
  qualitySnapshot(ids?: readonly string[]): Promise<NodeQualitySnapshot> { return this.#call("nodeQualitySnapshot", ids); }
  /** Read cumulative node statistics in `running` or `complete`, before ending the run.
   * @param ids - Ordered IDs; omit for all or use [] for none. Unknown/duplicate IDs reject.
   * @returns Detached statistics columns aligned with canonical `objectIds`.
   */
  statisticsSnapshot(ids?: readonly string[]): Promise<NodeStatisticsSnapshot> { return this.#call("nodeStatisticsSnapshot", ids); }
}
/** Owner-bound link handles. Snapshot selections preserve requested order; omit IDs for all links. */
export class LinkCollection extends ObjectCollection<Link> {
  readonly #call: Call;
  private constructor(ids: readonly string[], call: Call, assertOpen: () => void) {
    super(ids, (id) => Link.create(id, call), assertOpen); this.#call = call;
  }
  /** @internal */
  static create(ids: readonly string[], call: Call, assertOpen: () => void): LinkCollection { return new LinkCollection(ids, call, assertOpen); }
  /** Read link hydraulics in `running`, `complete`, or `ended`.
   * @param ids - Ordered IDs; omit for all or use [] for none. Unknown/duplicate IDs reject, including case-only duplicates.
   * @returns Detached hydraulic columns aligned with canonical `objectIds`, without a timestamp.
   */
  snapshot(ids?: readonly string[]): Promise<LinkSnapshot> { return this.#call("links", ids); }
  /** Read link quality in `running`, `complete`, or `ended`.
   * @param ids - Ordered IDs; omit for all or use [] for none. Unknown/duplicate IDs reject.
   * @returns Detached pollutant-major matrices; inner rows follow `objectIds`.
   */
  qualitySnapshot(ids?: readonly string[]): Promise<LinkQualitySnapshot> { return this.#call("linkQualitySnapshot", ids); }
  /** Read cumulative link statistics in `running` or `complete`.
   * @param ids - Ordered IDs; omit for all or use [] for none. Unknown/duplicate IDs reject.
   * @returns Detached scalar columns; flow-class durations have one seven-value row per link.
   */
  statisticsSnapshot(ids?: readonly string[]): Promise<LinkStatisticsSnapshot> { return this.#call("linkStatisticsSnapshot", ids); }
}
/** Owner-bound subcatchment handles and aligned hydraulic, quality, and statistics reads. */
export class SubcatchmentCollection extends ObjectCollection<Subcatchment> {
  readonly #call: Call;
  private constructor(ids: readonly string[], call: Call, assertOpen: () => void) {
    super(ids, (id) => Subcatchment.create(id, call), assertOpen); this.#call = call;
  }
  /** @internal */
  static create(ids: readonly string[], call: Call, assertOpen: () => void): SubcatchmentCollection { return new SubcatchmentCollection(ids, call, assertOpen); }
  /** Read runoff results in `running`, `complete`, or `ended`.
   * @param ids - Ordered IDs; omit for all or use [] for none. Unknown/duplicate IDs reject, including case-only duplicates.
   * @returns Detached runoff columns aligned with canonical `objectIds`, without a timestamp.
   */
  snapshot(ids?: readonly string[]): Promise<SubcatchmentSnapshot> { return this.#call("subcatchments", ids); }
  /** Read subcatchment quality in `running`, `complete`, or `ended`.
   * @param ids - Ordered IDs; omit for all or use [] for none. Unknown/duplicate IDs reject.
   * @returns Detached pollutant-major matrices; inner rows follow `objectIds`.
   */
  qualitySnapshot(ids?: readonly string[]): Promise<SubcatchmentQualitySnapshot> { return this.#call("subcatchmentQualitySnapshot", ids); }
  /** Read cumulative runoff statistics in `running` or `complete`.
   * @param ids - Ordered IDs; omit for all or use [] for none. Unknown/duplicate IDs reject.
   * @returns Detached statistics columns aligned with canonical `objectIds`.
   */
  statisticsSnapshot(ids?: readonly string[]): Promise<SubcatchmentStatisticsSnapshot> { return this.#call("subcatchmentStatisticsSnapshot", ids); }
}
/** Owner-bound rain-gage handles. Rain gages provide per-object results, not collection snapshots. */
export class RainGageCollection extends ObjectCollection<RainGage> {
  private constructor(ids: readonly string[], call: Call, assertOpen: () => void) {
    super(ids, (id) => RainGage.create(id, call), assertOpen);
  }
  /** @internal */
  static create(ids: readonly string[], call: Call, assertOpen: () => void): RainGageCollection { return new RainGageCollection(ids, call, assertOpen); }
}
